viewing_party/party.py

```diff
@@ -1,23 +1,173 @@
+
+    return user_data
+
 
-# -----------------------------------------
 # ------------- WAVE 2 --------------------
-# -----------------------------------------
+
+def get_watched_avg_rating(user_data):
+
+    ratings_list = []
+    
+    if len(user_data["watched"]) > 0:
+        for dict in user_data["watched"]:
+            ratings_list.append(dict["rating"])
+        
+        average_rating = sum(ratings_list) / len(ratings_list)
+        
+    else:
+        average_rating = 0.0
+    
+    return average_rating
+    
+    
+def get_most_watched_genre(user_data):
+
+    genre_frequency = {}
+
+
+    if len(user_data["watched"]) > 0:
+        for dict in user_data["watched"]:
+            for k, v, in dict.items():
+                if k == "genre":
+                    if v not in genre_frequency:
+                        genre_frequency[v] = 1
+                    else:
+                        genre_frequency[v] += 1
+
+ 
+        most_frequent_genre = max(genre_frequency, key=genre_frequency.get)
+    
+    else:
+        most_frequent_genre = None
+
+    return most_frequent_genre
 
 
-# -----------------------------------------
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
 
+def get_unique_watched(user_data):
+
+    friend_watched = []
+    difference = []
+
+    for dict in user_data["friends"]:
+        for k, v, in dict.items():
+                if k == "watched":
+                    friend_watched.extend(v)
+        
+    for dict in user_data["watched"]:
+        if dict not in friend_watched:
+            difference.append(dict)
+
+    return difference
+
+
+def get_friends_unique_watched(user_data):
+
+    friend_watched = []
+    difference1 = []
+    difference2 = []
+
+    for dict in user_data["friends"]:
+        for k, v, in dict.items():
+                if k == "watched":
+                    friend_watched.extend(v)
         
-# -----------------------------------------
+    for dict in friend_watched:
+        if dict not in user_data["watched"]:
+            difference1.append(dict)
+
+    for i in difference1:
+        if i not in difference2:
+            difference2.append(i)
+    
+    return difference2
+
+
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
+def get_available_recs(user_data):
+
+    reco_movies = []
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+
+    for dict in friends_unique_watched:
+        if dict["host"] in user_data["subscriptions"]:
+            reco_movies.append(dict)
+
+    return reco_movies
+
+
 # ------------- WAVE 5 --------------------
-# -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+
+    most_watched_genre = get_most_watched_genre(user_data)
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    reco_by_genre = []
+
+    for dict in friends_unique_watched:

```

Comment:
I've noticed `dict` is generally the default variable name used for dictionaries across this file. I highly suggest practicing to get out of this habit by using names that describe what the data represents. In most of cases through this file the `dict` objects represent movies, so I would call them as such or `movie_dict` if you want to be really clear on the data type.

Reply:
Thanks! I was having a hard time thinking about how deeply nested the data was and what I was actually accessing. The "dict" helped me to remember what data I was accessing. Will make a note to do something like "movie_dict" instead.